Python api development

JonathanESantos opened the pull request:
<!-- 
     All pull requests should fix a Triage Approved issue. Put that issue # here:
     e.g. ## Fixes #300. 
-->
## Fixes #500 
See also [Python support](https://feedback.azure.com/forums/920053-azure-kinect-dk/suggestions/37675798-python-support).

### Description of the changes:
- Adding Python k4a package with similar API as C++ and C# wrappers using ctypes as bindings to the C API. The bindings are for the API in the SDK v1.4.1.
- Adding functional tests to test Python k4a package.
- Example code is provided as a guide for using the Python k4a package.
- Build scripts is provided to package into wheel file which can be released as part of the SDK.
- Doxygen creates html documentation of the Python API. It is run as part of the build scripts.
- Documentation is provided for testing, building, and running the example code.

<!-- Please check off the appropriate boxes with [x] before submitting your pull request -->
### Before submitting a Pull Request:
- [x] I reviewed [CONTRIBUTING.md](https://github.com/Microsoft/Azure-Kinect-Sensor-SDK/blob/develop/CONTRIBUTING.md)
- [x] I [built my changes](https://github.com/Microsoft/Azure-Kinect-Sensor-SDK/blob/develop/docs/building.md) locally
- [x] I ran the [unit tests](https://github.com/Microsoft/Azure-Kinect-Sensor-SDK/blob/develop/docs/testing.md)
- [x] I ran the [functional tests](https://github.com/Microsoft/Azure-Kinect-Sensor-SDK/blob/develop/docs/testing.md) with a device
The following tests FAILED: 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 98, 99, 100, 101, 145, 146, 160, 161, 162.
- [x] I ran the [performance tests](https://github.com/Microsoft/Azure-Kinect-Sensor-SDK/blob/develop/docs/testing.md) with a device
"No tests were found!!!"

### I tested changes on: <!-- it's not required to have tested both, just indicate which one you tried -->
- [x] Windows
- [x] Linux


<!-- Specify how you tested your changes (i.e. manual/ad-hoc testing, automated testing, new automated tests added)-->
The additions are tested with new functional tests. The functional tests are not yet automated but they are easy to run in the command line or IDE. A document is provided for instructions on how to run the tests. The Python examples that use the API are also run.

This library has been tested with Python 3.8 and 3.9, but it should work for Python >= 3.6.

JonathanESantos committed:
PythonK4A: Adding initial directory structure, k4atypes.py which defines enums, and tests to check the enum values.

The _bindings/ folder is meant to contain internal ctypes bindings to the C-exported functions in k4a.dll.
The _libs/ folder is meant to contain the required DLLs: k4a.dll and DepthEngine.dll.
The api/ folder is meant for modules that define the actual Python K4A API.
The docs/ folder is for documentation.
The tests/ folder is for unit tests and functional tests.
The examples/ folder is for simple example code on how to use the API.

JonathanESantos committed:
PythonK4A: Updating __init__.py to point to the "api" directory (renamed from sdk).

JonathanESantos committed:
Python K4A: Rearranging directory tree so that it matches the expectation of what gets imported when users use "import k4a".
Adding a powershell script to build a wheel file. Adding empty docs files.

JonathanESantos committed:
Python k4a api: Put ctypes structs inside k4atypes.py.

Arrange import methods so that:
1. "import k4a" will expose all visible symbols in the module k4atypes.py.
2. "from k4a import *" will work the same way as "import k4a".

To see all symbols in k4atypes.py, use "import k4a.k4atypes".
This is considered to be a development or debug way of importing symbols.
Users are encouraged to use "import k4a".

JonathanESantos committed:
Python k4a api: Adding to setup.py so that "pip install -e .[test]" will install pytest along with k4a (in editable mode).

JonathanESantos committed:
Python k4a api: An enum name with AUTO was mistakenly replaced with _auto. Switching it back to AUTO.

For the enum tests, explicitly testing the enum entries rather than doing it in a for loop. These don't really test much other than making sure the enums have values from 0 to n.

JonathanESantos committed:
Python k4a api: Adding __repr__() functions to structs in order to print the contents of the struct.

JonathanESantos committed:
Python k4a api: Adding a check that the required dlls are in the _libs folder before making the wheel.

JonathanESantos committed:
Python k4a api: Adding the rest of the bindings for the functions that are exposed in the dll.

JonathanESantos committed:
Python k4a: Adding tests to call the functions in the dll.

TODO: Test the following functions:
k4a_calibration_3d_to_3d
k4a_calibration_2d_to_3d
k4a_calibration_3d_to_2d
k4a_calibration_2d_to_2d
k4a_calibration_color_2d_to_depth_2d
k4a_transformation_create
k4a_transformation_destroy
k4a_transformation_depth_image_to_color_camera
k4a_transformation_depth_image_to_color_camera_custom
k4a_transformation_color_image_to_depth_camera
k4a_transformation_depth_image_to_point_cloud

JonathanESantos committed:
Python k4a: Adding missing enum in color resolution, and adding tests for the transformation functions. The build_wheel.ps1 script will create a win-amd64 only python wheel.

JonathanESantos committed:
Python k4a: Changing @unittest.skip to @unittest.expectedfail for those tests that fail.

JonathanESantos committed:
Python k4a: Updating docs. Removing platform tag in building the k4a library wheel file.

JonathanESantos committed:
Python API: Renaming enums and structs to Python naming convention.

JonathanESantos committed:
Python API: Moving all internal files and symbols into the _bindings/ folder. By doing this, when a user "import k4a", only the exported symbols will be visible. Everything else will be under a _bindings symbol.

JonathanESantos committed:
Python API: Moving handle structs from _k4a.py back to _k4atypes.py so that it aligns more with k4a.h and k4atypes.h.

JonathanESantos committed:
Renaming modules inside _bindings subpackage to not include underscore. This seems redundant since the subpackage _bindings already has an underscore.

JonathanESantos committed:
Python api: Adding Device class which corresponds to a device with a device_handle.

JonathanESantos committed:
Python API: Adding capture and image classes to encapsulate capture and image functions and handles.
Adding test files.

JonathanESantos committed:
Python API: Bug in Image class was missing a height_pixels in one line, causing import k4a to fail.

JonathanESantos committed:
Python API: Adding Calibration and Transformation wrapper classes.

JonathanESantos committed:
Python API: Renaming a test file.

JonathanESantos committed:
Python API: Adding more tests to test the Device, Capture, Image, and Calibration classes. Fixing shallow copy and deep copy of Image objects.

JonathanESantos committed:
Python API: Fixing transformation functions and their tests.

JonathanESantos committed:
Python API: Adding example code that makes use of the Python API.
Fixing minor bugs in the Transformation class, and fixing issues with the unit tests regarding starting and stopping the imu and getting imu samples.

JonathanESantos committed:
Python API: Adding docstrings *.py files and a Doxygen settings file to generate the html documentation for the Python API.

JonathanESantos committed:
Python API: Adding a main page to the documentation, and fixing a bug in the example code image_transformations.py.

JonathanESantos committed:
Python API: Adding lines in build_wheel.ps1 to generate the documentation using doxygen.

Fixing several bugs:
1. The build_wheel.ps1 had extra arguments that breaks the building of the wheel.
2. The Transformation class had some parameters renamed but the renamed vars were not changed in the functions accordingly.
3. The Transformation class removed an unneeded param in depth_image_to_color_camera_custom but did not remove it in the test code.

JonathanESantos committed:
Making minor changes to get this k4a package to work in both Windows and Linux. Adding a bash script to automate generating the wheel file in Ubuntu (Linux).

JonathanESantos committed:
Forgot to commit the minor changes to get this k4a package to work in both Windows and Linux. Doing it now.

JonathanESantos committed:
Python API: Specifying specific library names that the python code will try to load. In Windows, the library names must be k4a.dll and depthengine.dll. In Linux, the library names must be libk4a.so and libdepthengine.so.

If the libraries have different names, the user should create symlinks and point it to the actual library using relative addressing.

JonathanESantos committed:
Python API: Modifying the build_wheel.csh script to look for libdepthengine.so and libk4a.so file names.

JonathanESantos committed:
Python API: Updating building.md for clarity.

ghost commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/microsoft/Azure-Kinect-Sensor-SDK?pullRequest=1483) <br/>All CLA requirements met.

JonathanESantos committed:
Python API: Fixing bug in tests that causes failure in some of the tests. Also renaming the tests so that they can be filtered into unit, functional_fast, functional, and perf tests.

lijerryli commented on src/python/k4a/README.md:
```diff
@@ -0,0 +1,77 @@
+# Azure Kinect Python API(K4A)
```
test comment

lijerryli reviewed (commented):
testing to make sure this works

lijerryli commented on src/python/k4a/docs/building.md:
```diff
@@ -0,0 +1,123 @@
+
+OS:
+* Windows
+* Linux
```
Is there a specific version of Linux that is required (like Ubuntu 18.04)

qm13 replied:
Sensor SDK currently supports Ubuntu 18.04 only. The body tracking team has verified it also supports Ubuntu 20.04 (just released Body Tracking v1.1.0) and we should update Sensor SDK to support both Ubuntu 18.04 and 20.04. Consequently we should test the python wrapper on both Ubuntu 18.04 and 20.04.

JonathanESantos replied:
I've actually tested the python wrapper on Ubuntu 20.04 "informally". It seems to work.

lijerryli commented on src/python/k4a/docs/building.md:
```diff
@@ -0,0 +1,123 @@
+  
+* [k4a library](../../../../docs/building.md)
+  The k4a library can be built from the repository or the SDK can be downloaded
+  for the binary. The Windows library is k4a.dll, and the Linux library is k4a.so.
```
The Linux library is called libk4a.so

qm13 replied:
Please fix.

JonathanESantos replied:
Changed to libk4a.so.

lijerryli commented on src/python/k4a/docs/building.md:
```diff
@@ -0,0 +1,123 @@
+   
+   Note: To create a cross-platform wheel, the .dll for Windows and the .so for Linux need to
+         copied into the folder.
+   Note: If the dlls have different names, create a symlink with relative redirection. This may
```
Add a \ for a new line, right now these notes all get thrown together

qm13 replied:
Please fix.

JonathanESantos replied:
Fixed these formatting issues here and in all the other markdowns.

lijerryli commented on src/python/k4a/docs/building.md:
```diff
@@ -0,0 +1,123 @@
+   
+   Note: To create a cross-platform wheel, the .dll for Windows and the .so for Linux need to
+         copied into the folder.
+   Note: If the dlls have different names, create a symlink with relative redirection. This may
```
Add a \ for a new line, right now these notes all get thrown together

qm13 replied:
Please fix.

JonathanESantos replied:
Fixed these formatting issues here and in all the other markdowns.

lijerryli commented on src/python/k4a/docs/building.md:
```diff
@@ -0,0 +1,123 @@
+The following dependencies are for both Windows and Linux:
+
+* [python3](https://www.python.org/getit/). During the install make sure to add
+  python to path.
```
Also need venv

lijerryli replied:
as well as pip and doxygen, but on my system I only needed to manually install venv additionally

qm13 replied:
Please fix.

JonathanESantos replied:
I did notice that venv is not automatically installed as a Python package in Linux, even though venv is part of the Python core modules. It is part of Python on Windows...In any case, I added a note to install venv in Linux.

There is a comment that doxygen is an optional tool.

As for pip, it should really be installed as part of Python. It is the official Python package installer, so yeah. If it is not part of Python installation, then nothing will work. In any case, I added a comment about it.

lijerryli commented on src/python/k4a/docs/building.md:
```diff
@@ -0,0 +1,123 @@
+   
+   Note: To create a cross-platform wheel, the .dll for Windows and the .so for Linux need to
+         copied into the folder.
+   Note: If the dlls have different names, create a symlink with relative redirection. This may
```
Add a \ for a new line, right now these notes all get thrown together

lijerryli commented on src/python/k4a/docs/examples.md:
```diff
@@ -0,0 +1,65 @@
+To run, open a command terminal and type: python the_basics.py
+The expected output is a print out on the terminal of the device, capture, and image properties.
+
+### image_transformations.py
```
both image_transformations and simple_viewer are using plt.show() commands, which requires a GUI component. I had to install tkinter using `sudo apt install python3-tk` for this to work

JonathanESantos replied:
Added tkinter as a requirement.

lijerryli commented on src/python/k4a/docs/examples.md:
```diff
@@ -0,0 +1,65 @@
+
+A simple program to showcase the image transformation functions in the Azure Kinect API.
+
+Additional Prerequisites: 
```
newline using \\
currently all on one line

lijerryli commented on src/python/k4a/src/k4a/_bindings/k4a.py:
```diff
@@ -0,0 +1,525 @@
+
+#K4A_EXPORT size_t k4a_image_get_size(k4a_image_t image_handle);
+k4a_image_get_size = _k4a_lib.k4a_image_get_size
+k4a_image_get_size.restype = _ctypes.c_ulonglong
```
there is a ctpyes c_size_t
That might be more extensible if we want to support both 32 and 64 bit

lijerryli commented on src/python/k4a/src/k4a/_bindings/k4a.py:
```diff
@@ -0,0 +1,525 @@
+k4a_image_create_from_buffer.restype = EStatus
+k4a_image_create_from_buffer.argtypes=(
+    _ctypes.c_int, _ctypes.c_int, _ctypes.c_int, _ctypes.c_int, _ctypes.POINTER(_ctypes.c_uint8),
+    _ctypes.c_ulonglong, _memory_allocate_cb, _ctypes.c_void_p, _ctypes.POINTER(_ImageHandle))
```
there is a ctpyes c_size_t
That might be more extensible if we want to support both 32 and 64 bit

JonathanESantos replied:
Changed to c_size_t. Thanks for letting me know that!

lijerryli commented on src/python/k4a/src/k4a/_bindings/k4a.py:
```diff
@@ -0,0 +1,525 @@
+k4a_device_get_serialnum = _k4a_lib.k4a_device_get_serialnum
+k4a_device_get_serialnum.restype = EBufferStatus
+k4a_device_get_serialnum.argtypes = (_DeviceHandle, 
+    _ctypes.POINTER(_ctypes.c_char), _ctypes.POINTER(_ctypes.c_ulonglong))
```
there is a ctpyes c_size_t
That might be more extensible if we want to support both 32 and 64 bit

JonathanESantos replied:
Changed to c_size_t.

lijerryli commented on src/python/k4a/src/k4a/_bindings/k4a.py:
```diff
@@ -0,0 +1,525 @@
+#                                                              size_t *data_size);
+k4a_device_get_raw_calibration = _k4a_lib.k4a_device_get_raw_calibration
+k4a_device_get_raw_calibration.restype = EBufferStatus
+k4a_device_get_raw_calibration.argtypes = (_DeviceHandle, _ctypes.POINTER(_ctypes.c_uint8), _ctypes.POINTER(_ctypes.c_ulonglong))
```
there is a ctpyes c_size_t
That might be more extensible if we want to support both 32 and 64 bit

JonathanESantos replied:
Changed to c_size_t.

lijerryli commented on src/python/k4a/src/k4a/_bindings/k4a.py:
```diff
@@ -0,0 +1,525 @@
+#                                                     k4a_calibration_t *calibration);
+k4a_calibration_get_from_raw = _k4a_lib.k4a_calibration_get_from_raw
+k4a_calibration_get_from_raw.restype = EStatus
+k4a_calibration_get_from_raw.argtypes = (_ctypes.POINTER(_ctypes.c_char), 
```
there is a ctpyes c_size_t
That might be more extensible if we want to support both 32 and 64 bit

JonathanESantos replied:
Changed to c_size_t.

lijerryli commented on src/python/k4a/docs/testing.md:
```diff
@@ -0,0 +1,119 @@
+      
+To capture the results, use an additional --junit-xml=./test_results.xml option.
+
+### Running performance tests on the command line
```
There doesn't seem to be any perf tests right now? Maybe add a comment saying so

qm13 replied:
Ok

JonathanESantos replied:
Added comment about there not being performance tests right now.

kyrendon-msft commented on src/python/k4a/docs/building.md:
```diff
@@ -0,0 +1,123 @@
+
+The following tools are optional:
+
+* [doxygen](https://github.com/doxygen/doxygen)
```
Build scripts fail if doxygen is not installed, but it is listed here as optional.

qm13 replied:
Doxygen is required to generate the API Reference documentation.

JonathanESantos replied:
Added check to test if doxygen is installed so that it is indeed optional.

JonathanESantos committed:
Removing forced-loading of depthengine library. Just load k4a.dll and let it load the depth engine.

Fixing markdown formatting.

JonathanESantos committed:
In the wheel build scripts, adding test to check if doxygen is installed.
In the building.md docs, adding a Linux requirement to install venv. For some reason, it is not installed natively as part of a Python installation.

JonathanESantos committed:
Fixing markdown documents.

JonathanESantos committed:
Adding blurb about installing pip in Linux.

JonathanESantos committed:
Changing ctypes.c_ulonglong to ctypes.c_size_t where it matches the function signature on the C API better.

JonathanESantos committed:
Changing c_ulonglong to c_size_t in the tests to match the change in the bindings.

JonathanESantos committed:
Fixing build script in Linux to look for depthengine.so* inside a subfolder. Fixing docs to reflect these changes.

JonathanESantos committed:
Fixing documents that say the Linux depthengine.so.2.0 must be in a subfolder libk4a1.4. This is a previous method of chain loading the libraries, which is not the case anymore. The depthengine.so.2.0 should be in the same folder as the libk4a.so.

jaygullapalli commented on src/python/k4a/docs/building.md:
```diff
@@ -79,14 +79,14 @@ The following tools are optional:
    folder named libk4a\<ver\>, where \<ver\> is the version major.minor of the k4a library
    (e.g. _libs constains libk4a1.4/libdepthengine.so.2.0).
```
This comment ("The depth engine library must be in a folder named libk4a\<ver> ... ") is no longer relevant due to the changes, right?

JonathanESantos replied:
Correct. Removed it. Thanks for catching.

JonathanESantos committed:
The build script for Linux needs to check for depthengine.so* in the same folder as libk4a.so.

jaygullapalli commented on src/python/k4a/docs/building.md:
```diff
@@ -79,14 +79,14 @@ The following tools are optional:
+    >Python will look for "libk4a.so" in Linux which will also load the depth engine dll.
 
     >**Note:** An example of the files in _libs for Linux are: 
     >- libk4a.so, a soft link to libk4a.1.4
```
These are the files on Linux (looks like we are missing file suffixes in the instructions)
libk4a.so (link) -> libk4a.so.1.4
libk4a.so.1.4 (link) -> libk4a.so.1.4.0
libk4a.so.1.4.0

JonathanESantos replied:
I modified the wording slightly in the most recent commits. See if it's clearer.

JonathanESantos committed:
Making slight changes to the building.md documentation for more clarity.

jaygullapalli commented on src/python/k4a/examples/image_transformations.py:
```diff
@@ -0,0 +1,110 @@
+'''
+image_tarnsformations.py
```
Typo in transformations

JonathanESantos replied:
Fixed in latest update.

JonathanESantos committed:
Fixing a minor typo in the comments of image_transformations.py, and adding additional comments to let the users know they have to install a required package separately from k4a package.
